org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/corext/template/java/JavaContextType.java

```diff
@@ -58,4 +77,74 @@ protected void initializeContext(IJavaContext context) {
 			context.addCompatibleContextType(JavaContextType.ID_ALL);
 		}
 	}
+
+	/* (non-Javadoc)
+	 * @see org.eclipse.jdt.internal.corext.template.java.AbstractJavaContextTypeCore#initializeContextTypeResolvers()
+	 */
+	@Override
+	public void initializeContextTypeResolvers() {
+		super.initializeContextTypeResolvers();
+		// TODO: Some of the resolvers are defined in org.eclipse.jdt.ui/plugin.xml, now we have to add them manually.
+		// See: https://github.com/eclipse/eclipse.jdt.ui/blob/cf6c42522ee5a5ea21a34fcfdecf3504d4750a04/org.eclipse.jdt.ui/plugin.xml#L5619-L5625
+		addResolver("var", new VarResolver());
+		addResolver("localVar", new LocalVarResolver());
+		addResolver("formattedLocalVar", new FormattedLocalVarResolver());
+		addResolver("enumerated_method_arguments", new EnumeratedMethodArgumentsResolver());
+	}
+
+	public synchronized void addResolver(String type, TemplateVariableResolver resolver) {
+		resolver.setType(type);
+		addResolver(resolver);
+	}
+
+	/**
+	 * Resolves ${enumerated_method_arguments} variable into:
+	 *
+	 * <pre>
+	 * "arg1="+arg1+", arg2="+arg2+...+", argN="+argN
+	 * </pre>
+	 */
+	static class EnumeratedMethodArgumentsResolver extends TemplateVariableResolver {
+
+		@Override
+		protected String resolve(TemplateContext context) {
+			if (!(context instanceof CompilationUnitContext)) {
+				return null;
+			}
+			IJavaElement element = ((CompilationUnitContext) context).findEnclosingElement(IJavaElement.METHOD);
+			if (element != null) {
+				try {
+					return ArgumentsHelper.format(((IMethod) element).getParameterNames());
+				} catch (Exception O_o) {
+					// ¯\_(ツ)_/¯
+				}
+			}
+			return null;
+		}
+	}
+
+	static class FormattedLocalVarResolver extends TemplateVariableResolver {
+
+		@Override
+		public String resolve(TemplateContext context) {
+			String[] latestVariable = new String[1];
+			if (context instanceof JavaContextCore) {
+				JavaContextCore jc = (JavaContextCore) context;
+				try {
+					jc.getCompilationUnit().codeComplete(jc.getStart(), new CompletionRequestor() {
```

Comment:
it calls codeComplete twice, is there any performance impact?

Reply:
It is indeed a good question. Here's the thing, we could avoid using this custom resolver and use the following snippet instead, which relies on the existing LocalVarResolver class:
> System.out.println("${localVar} ="+${localVar});

But:
- when there are no local variables, it literally outputs `System.out.println("localVar ="+ localVar);`, (IJ outputs `System.out.println("true = "+ true)` which is barely any better)
- primitive local vars are not inserted (WTF !?!)

So, when looking at how LocalVarResolver computes the local var, you can see a call to JavaContextCore.getLocalVariables:

```java
@Override
public Variable[] getLocalVariables(String type) {
    Variable[] localVariables= getCompletion().findLocalVariables(type);
    arrange(localVariables);
    return localVariables;
}
```

digging deeper into getCompletion():

```java
private CompilationUnitCompletion getCompletion() {
    ICompilationUnit compilationUnit= getCompilationUnit();
    if (fCompletion == null) {
        fCompletion= new CompilationUnitCompletion(compilationUnit);

        if (compilationUnit != null) {
            try {
                compilationUnit.codeComplete(getStart(), fCompletion);
            } catch (JavaModelException e) {
                // ignore
            }
        }
    }

    return fCompletion;
}
```
So the code completion call is there too, except that, its result is cached into the common JavaContext, for the benefit of other resolvers being used in the same completion request.
The main problem then, is CompilationUnitCompletion.findLocalVariables doesn't return any primitive values, which seems absurd:

```java
public Variable[] findLocalVariables(String clazz) {
    List<Variable> matches= new ArrayList<>();

    for (ListIterator<Variable> iterator= fLocalVariables.listIterator(fLocalVariables.size()); iterator.hasPrevious();) {
        Variable localVariable= iterator.previous();

         // THIS CHECK HERE PREVENTS COLLECTION OF PRIMITIVE VARIABLES
        if (localVariable.isSubtypeOf(clazz))
            matches.add(localVariable);
    }

    return matches.toArray(new Variable[matches.size()]);
}
```

So the big drawback of this variable resolver is that, code completion can be called 3 times:
- once by the user
- once by JavaContext, if we're matching snippets that include some var or localVar placeholder
- once by my naive FormattedLocalVarResolver implementation to compute actual local variables

This is definitely not ideal

Reply:
An awesome finding! When i analyzed code completion performance earlier, i also found computing snippet proposals took much time and has a big impact on the total completion time. This explained it.

// @Eskibear might be interested too.